drivers/i2s/i2s_nrfx.c

```diff
@@ -0,0 +1,1328 @@
+		int ret = interface_start(i2s);
+
+		if (ret < 0) {
+			interface_error_service(i2s, "Internal ISR error");
+		}
+		return;
+	}
+}
+
+#define I2S_NRFX_DEVICE(idx)						       \
+static struct queue_item q_rx_##idx##_buf[CONFIG_NRFX_I2S_RX_BLOCK_COUNT + 1]; \
+static struct queue_item q_tx_##idx##_buf[CONFIG_NRFX_I2S_TX_BLOCK_COUNT + 1]; \
+static void i2s_nrfx_irq_##idx##_config(struct device *dev);		       \
+									       \
+static void setup_instance_##idx(struct device *dev)			       \
+{									       \
+	struct i2s_nrfx_interface *i2s = get_interface(dev);		       \
+									       \
+	queue_init(&i2s->channel_tx.mem_block_queue,			       \
+		CONFIG_NRFX_I2S_TX_BLOCK_COUNT + 1, &q_tx_##idx##_buf[0]);     \
+	queue_init(&i2s->channel_rx.mem_block_queue,			       \
+		CONFIG_NRFX_I2S_RX_BLOCK_COUNT + 1, &q_rx_##idx##_buf[0]);     \
+	i2s_nrfx_irq_##idx##_config(dev);				       \
+}									       \
+									       \
+static const struct i2s_nrfx_config channel_cfg_##idx = {		       \
+	.sck_pin = DT_NORDIC_NRF_I2S_I2S_##idx##_SCK_PIN,		       \
+	.lrck_pin = DT_NORDIC_NRF_I2S_I2S_##idx##_LRCK_PIN,		       \
+	.mck_pin = DT_NORDIC_NRF_I2S_I2S_##idx##_MCK_PIN,		       \
+	.sdout_pin = DT_NORDIC_NRF_I2S_I2S_##idx##_SDOUT_PIN,		       \
+	.sdin_pin = DT_NORDIC_NRF_I2S_I2S_##idx##_SDIN_PIN,		       \
+	.instance_init = setup_instance_##idx,				       \
+};									       \
+									       \
+static struct i2s_nrfx_data channels_data_##idx = {			       \
+	.interface = {							       \
+		.state = I2S_IF_NOT_READY,				       \
+		.size = 0,						       \
+		.channel_tx = {						       \
+			.current_state = I2S_STATE_NOT_READY,		       \
+			.last_trigger_cmd = I2S_TRIGGER_PREPARE,	       \
+			.mng = &ch_mng.tx,				       \
+		},							       \
+		.channel_rx = {						       \
+			.current_state = I2S_STATE_NOT_READY,		       \
+			.last_trigger_cmd = I2S_TRIGGER_PREPARE,	       \
+			.mng = &ch_mng.rx,				       \
+		}							       \
+	}								       \
+};									       \
+									       \
+DEVICE_AND_API_INIT(i2s_##idx, DT_NORDIC_NRF_I2S_I2S_##idx##_LABEL,	       \
+		    &i2s_nrfx_initialize, &channels_data_##idx,		       \
+		    &channel_cfg_##idx, POST_KERNEL,			       \
+		    CONFIG_I2S_INIT_PRIORITY, &i2s_nrf_driver_api);	       \
+									       \
+static void i2s_nrfx_irq_##idx##_config(struct device *dev)		       \
+{									       \
+	IRQ_CONNECT(DT_INST_##idx##_NORDIC_NRF_I2S_IRQ_0,		       \
+		DT_INST_##idx##_NORDIC_NRF_I2S_IRQ_0_PRIORITY,		       \
+		isr, DEVICE_GET(i2s_##idx), 0);				       \
+	irq_enable(DT_INST_##idx##_NORDIC_NRF_I2S_IRQ_0);		       \
+}
+
+static const struct i2s_driver_api i2s_nrf_driver_api = {
+	.configure = i2s_nrfx_api_configure,
+	.read = i2s_nrfx_read,
+	.write = i2s_nrfx_write,
+	.trigger = i2s_nrfx_trigger,
+};
+
+struct channels_management {
+	struct dir_mng rx;
+	struct dir_mng tx;
+};
+
+static const struct channels_management ch_mng = {
+	.rx = {
+		.start = channel_rx_start,
```

Comment:
I still don't see the advantage of using these function pointers. This way you introduce extra overhead in each call of these functions, and you effectively prevent compiler from doing any optimizations in their calling, while some of them, like `_stop` or `_drain`, are likely to be inlined.
And I could not find a place in the current implementation in which you could not easily replace the pointer dereferencing by a direct call of the corresponding function.